clippy_lints/src/loops/mod.rs

```diff
@@ -635,6 +636,48 @@ declare_clippy_lint! {
     "checking for emptiness of a `Vec` in the loop condition and popping an element in the body"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for infinite loops in a function where the return type is not `!`
+    /// and lint accordingly.
+    ///
+    /// ### Why is this bad?
+    /// A loop should be gently exited somewhere, or at lease mark its parent function as
+    /// never return (`!`).
+    ///
+    /// ### Example
+    /// ```no_run,ignore
+    /// fn run_forever() {
+    ///     loop {
+    ///         // do something
+    ///     }
+    /// }
+    /// ```
+    /// If infinite loops are as intended:
+    /// ```no_run,ignore
+    /// fn run_forever() -> ! {
+    ///     loop {
+    ///         // do something
+    ///     }
+    /// }
+    /// ```
+    /// Otherwise add a `break` or `return` condition:
+    /// ```no_run,ignore
+    /// fn run_forever() {
+    ///     loop {
+    ///         // do something
+    ///         if condition {
+    ///             break;
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    #[clippy::version = "1.75.0"]
+    pub INFINITE_LOOP,
+    restriction,
+    "possibly unintended infinite loops"
```

Comment:
I also wondering about weird fn like
````rust
fn loopfn() -> Result<(), i32> {
    Ok(loop {})
}
````

afaik something like ` -> Result <!, i32>` is not stabilized yet.

Reply:
> I also wondering about weird fn like

Hmmm, I wasn't aware such cases, currently it would just prints help message suggesting add a break condition...
What would be an ideal solution here :thinking: ?

Reply:
Happy holiday :smile: 

As of your concern, yes it will lint, right now the visitor only looks for following occurrence in a loop block:

1. A return or a break
2. A function which never return, which was inspired be one of your early suggestion such as the one in `panic!`, `unreachable!`, also `std::process::exit`.

And if non of these could be find then... well, triggers the warning.